### ISSUE TITLE
Stochastic behavior when extracting the features for one image

### ISSUE THREAD

**Issue by maduc7** [mahmoodlab/HIPT#32]
Hi,
Really great work!

I have been extracting features at the two first levels (256 and 4k) for one whole slide images. Running it four times (to see if it was reproducible), I noticed that I got four different features. Setting `torch.backends.cudnn.deterministic=True` solved the issue, however I still do not get why we are having differences as the weights should be fixed.

The line that is creating the stochastic behavior is the following: 
from line 43 in HIPT_4K/hipt_4k.py: `self.model256 = get_vit256(pretrained_weights=model256_path).to(device256)`
then line 54 in HIPT_4K/hipt_model_utils.py: `model256 = vits.__dict__[arch](patch_size=16, num_classes=0)` (which is calling vit_small)
then line 165 in: HIPT_4K/vision_transformer.py: `self.proj = nn.Conv2d(in_chans, embed_dim, kernel_size=patch_size, stride=patch_size)`

This last line is where I found the stochastic behavior. Why is that?
Thank

**Comment by Arkkienkeli** [comment]
Hi @maduc7, parallel computations can be the source of non-deterministic behavior, I have a little experience with PyTorch though, but [others are also complaining about that](https://stackoverflow.com/questions/72979303/why-is-pytorch-inference-non-deterministic-even-when-setting-model-eval). 
Never noticed that working with TensorFlow though (in version 2+). 
Out of curiosity: how much features differ between runs without `torch.backends.cudnn.deterministic=True` ? 
Maybe the maintainers can give a better hint!

**Comment by maduc7** [comment]
Thanks for your reply. I guess a slight fluctuation would not worry me that much but in that image, the features could differ up to 5 (absolute value). And the features range for that image are going from -4 to 4 so 5 is really a lot.

**Comment by Richarizardd** [comment]
Hi @maduc7 - apologies for the delay. I was looking into this issue to try and reproduce it to stochasticity, but did not see the variations that you reported. Some comments:
1. Are you running the model with `model.eval()` and `torch.no_grad() / torch.inference()`? The model output may be different due to dropout and other stochastic behavior.
2. I tested this issue across a variety of CUDA versions (10.1, 10.2, 11.2, 11.7) and torch versions (1.6.1, 1.8.1, 1.13.1, 2.0.0-dev). While the outputs were deterministic, I did find small subtleties between the outputs across different versions (varied in the range of -0.2 to 0.2 using `np.linalg.norm`), which is not as large as the range that you are seeing.

Let me know if you have more details about this issue, and I could look into it more.

**Comment by maduc7** [comment]
Hi @Richarizardd,

Thanks for your answer and for your feedbacks.
1. Yes, I used model.eval() and torch.inference_mode()
2. I used CUDA 11.6 and torch versions 1.13.0

For me what solved it was to set the seed for torch.manual_seed(seed) but it is still confusing to me to have stochasticity given the model should be set fixed.

**Comment by GzrDavidRicardo** [comment]
Hi @maduc7, may I ask did you insert the HIPT_4K API into the CLAM extract_feature_fp.py script for feature extraction? If so, could you kindly provide the package dependencies and their version? I kept encountering unexpected issues during feature extraction using HIPT 4K. Thank you.
